internal/graph/cached_resolver.go

```diff
@@ -0,0 +1,128 @@
+package graph
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/gob"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+const (
+	defaultMaxCacheSize = 10000
+	defaultCacheTTL     = 10 * time.Second
+)
+
+// CachedCheckResolver implements the CheckResolver interface in way that attempts to resolve
+// Check subproblems via prior computations before delegating the request to some underlying
+// CheckResolver.
+type CachedCheckResolver struct {
+	delegate     CheckResolver
+	cache        *ccache.Cache[*ResolveCheckResponse]
+	maxCacheSize int64
+	cacheTTL     time.Duration
+}
+
+var _ CheckResolver = (*CachedCheckResolver)(nil)
+
+// CachedCheckResolverOpt defines an option that can be used to change the behavior of cachedCeckResolver
+// instance.
+type CachedCheckResolverOpt func(*CachedCheckResolver)
+
+// WithMaxCacheSize sets the maximum size of the Check resolution cache. After this
+// maximum size is met, then cache keys will start being evicted with an LRU policy.
+func WithMaxCacheSize(size int64) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.maxCacheSize = size
+	}
+}
+
+// WithCacheTTL sets the TTL (as a duration) for any single Check cache key value.
+func WithCacheTTL(ttl time.Duration) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cacheTTL = ttl
+	}
+}
+
+func WithExistingCache(cache *ccache.Cache[*ResolveCheckResponse]) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cache = cache
+	}
+}
+
+// NewCachedCheckResolver constructs a CheckResolver that delegates Check resolution to the provided delegate,
+// but before delegating the query to the delegate a cache-key lookup is made to see if the Check subproblem
+// has already recently been computed. If the Check subproblem is in the cache, then the response is returned
+// immediately and no recomputation is necessary.
+func NewCachedCheckResolver(delegate CheckResolver, opts ...CachedCheckResolverOpt) *CachedCheckResolver {
+	checker := &CachedCheckResolver{
+		delegate:     delegate,
+		maxCacheSize: defaultMaxCacheSize,
+		cacheTTL:     defaultCacheTTL,
+	}
+	checker.cache = ccache.New(
+		ccache.Configure[*ResolveCheckResponse]().MaxSize(checker.maxCacheSize),
+	)
+
+	for _, opt := range opts {
+		opt(checker)
+	}
+
+	return checker
+}
+
+func (c *CachedCheckResolver) ResolveCheck(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+) (*ResolveCheckResponse, error) {
+
+	dispatchedCheckCounter.Inc()
+
+	cacheKey, err := checkRequestCacheKey(req)
+	if err != nil {
+		return nil, fmt.Errorf("cache key computation failed with error: %w", err)
```

Comment:
if we can't compute the cache key we should avoid the cache altogether, no?

Reply:
In theory we can avoid the cache.  However, in reality, I don't see any scenario where the cache key fails to be computed - there is something seriously weird and it is better to return error at this point.

Reply:
> Or some other option that doesn't surface this error to users but we still know about it..

This is not possible as OpenFGA team is not the operator.

Reply:
Updated just to swallow + log

Reply:
I actually don't agree with swallow + log in this case. If we cannot compute the cache key, then that indicates something much more severely wrong (maybe tuplekey validation is broken in our code etc..). If we fail to compute a cache key, then there is an invariant that is not being met and we should seriously fix it. Failure to fix the issue can continue to lead to bypassing the cache, which will degrade performance unnecessarily and void the benefit of the cache in the first place. In such cases, it truly is an internal error worth noting and causing server errors for. We want to strongly encourage/enforce an invariant that requires this never fail. For that reason, I want to raise awareness to an error such as this very quickly, and logging an error isn't going to drive us to fix the issue faster that the alternatives. I would vote that we return an internal error and log the underlying error. Second vote would be panic, though I feel that's a bit drastic in this case within the request lifecycle.

Reply:
If we were communicating with some remote cache server (e.g. Redis) and there was some transient network issue that caused an error, then I would agree that we should log the error and evaluate the request. But this cache key computation is on a fixed value that should have strong invariants met, and so it should be guaranteed. If it is not, then that is not a transient thing like networking issues to a Redis server, that's a severe bug.

Reply:
Change to return error. If it is transient error, let's have the caller retry.